packages/@functionless/formation/src/resource-provider.ts

```diff
@@ -0,0 +1,140 @@
+import { NoOpProvider } from "./resource-providers/no-op";
+import { EventSourceMappingProvider } from "./resource-providers/event-source-mapping";
+
+export interface CreateRequest<Properties> {
+  logicalId: string;
+  resourceType: ResourceType;
+  definition: Properties;
+}
+
+export interface UpdateRequest<Properties> extends CreateRequest<Properties> {
+  previous: PhysicalResource<Properties>;
+}
+
+export interface DeleteRequest<Properties> {
+  logicalId: string;
+  resourceType: ResourceType;
+  physicalId: string;
+  previous: PhysicalResource<Properties>;
+  /**
+   * True when the resource was already snapshot before the delete.
+   */
+  snapshotDone: boolean;
+}
+
+export interface ResourceOperationResultMetadata {
+  /**
+   * Minimum milliseconds to wait after all deployment operations are complete.
+   *
+   * If it takes around 10s for a Policy update (paddingMillis: 10000) to be reflected (consistency), but
+   * the rest of the deployment only takes 5s, we will wait at least 5s before completing the deployment.
+   *
+   * TODO: replace with a consistent vs referable API.
+   */
+  paddingMillis?: number;
+}
+
+export type ResourceOperationResult<Properties = any> = Promise<
+  | PhysicalResource<Properties>
+  | ({
+      resource: PhysicalResource<Properties>;
+    } & ResourceOperationResultMetadata)
+>;
+
+export interface ResourceProviderRetryConfig {
+  canRetry: ("CREATE" | "DELETE" | "UPDATE")[] | true;
+}
+
+/**
+ * TODO: support optional snapshot.
+ */
+export interface ResourceProvider<Properties = any> {
+  retry?: ResourceProviderRetryConfig;
+  create(
+    request: CreateRequest<Properties>
+  ): ResourceOperationResult<Properties>;
+  update(
+    request: UpdateRequest<Properties>
+  ): ResourceOperationResult<Properties>;
+  delete(
+    request: DeleteRequest<Properties>
+  ): Promise<void | ResourceOperationResultMetadata>;
+}
+
+export type ResourceProviderInitializer<Properties = any> = (
+  props: ResourceProviderProps
+) => ResourceProvider<Properties>;
+
+export const DEFAULT_RESOURCE_PROVIDER_KEY = "FORMLESS_DEFAULT";
+
+export const DefaultResourceProviders: Record<
+  string,
+  ResourceProvider | ResourceProviderInitializer
+> = {
+  "AWS::Events::EventBus": (props) => new EventBusProvider(props),
+  "AWS::Events::Rule": (props) => new EventBusRuleProvider(props),
+  "AWS::IAM::Policy": (props) => new InlinePolicyProvider(props),
+  "AWS::IAM::ManagedPolicy": (props) => new ManagedPolicyProvider(props),
+  "AWS::SQS::QueuePolicy": (props) => new QueuePolicyProvider(props),
+  "AWS::SQS::Queue": (props) => new QueueProvider(props),
+  "AWS::IAM::Role": (props) => new RoleProvider(props),
+  "AWS::SecretsManager::Secret": (props) => new SecretProvider(props),
+  "AWS::CDK::Metadata": new NoOpProvider(),
+  "AWS::Lambda::EventSourceMapping": (props) =>
+    new EventSourceMappingProvider(props),
+  [DEFAULT_RESOURCE_PROVIDER_KEY]: (props) => new CloudControlProvider(props),
```

Comment:
Would it a be better to have this logic handled explicitly by the caller? Strange to see it as a magic key

Reply:
these are defaults, the user can pass in any map of providers or merge into the defaults. 

What interface were you looking for?

Reply:
I would expect this map to only have resource types. What resource is the default? Is there logic to specially handle that key?

Reply:
Default is the provider we try when there is no explicit key for the current resource in the map. It seemed easier to define the default providers this way, at least until a more complex fallback system was created.

Other options: 
* a map of key to providers and a parameter for "defaultProvider"
* a pipeline approach where a single provider is given which can have providers nested within.
* a map of key to providers then an array of fallback providers.
* a single handler given by the consumer which returns the provider given a key and some config